management_app/views/course.py

```diff
@@ -101,74 +120,249 @@ def upload_user_file():
-            if num_of_enrollment is None:
-                teaching_point_val = 1  # default value is 1 for each course
-            else:
-                teaching_point_val = calculate_teaching_point_val(course_title_id, num_of_enrollment, offload_or_recall_flag, year, quarter)
-
-            # check existence
-            row = db.execute(
-                'SELECT * FROM scheduled_teaching '
-                ' WHERE user_id = ? AND year = ? AND quarter = ? AND course_title_id = ? AND course_sec = ?', 
-                (user_id, year, quarter, course_title_id, course_sec)
-            ).fetchone()
-            if row is None:
-                db.execute(
-                    'INSERT INTO scheduled_teaching (user_id, year, quarter, course_title_id, course_sec, enrollment, offload_or_recall_flag, teaching_point_val)'
-                    ' VALUES (?, ?, ?, ?, ?, ?, ?, ?)',
-                    (user_id, year, quarter, course_title_id, course_sec, num_of_enrollment, offload_or_recall_flag, teaching_point_val)
-                )
-            else:
-                db.execute(
-                    'UPDATE scheduled_teaching SET enrollment = ?, offload_or_recall_flag = ?, teaching_point_val = ?'
-                    ' WHERE user_id = ? AND year = ? AND quarter = ? AND course_title_id = ? AND course_sec = ?', 
-                    (num_of_enrollment, offload_or_recall_flag, teaching_point_val, user_id, year, quarter, course_title_id, course_sec)
-                )
-            db.commit()
+            num_of_enrollment = get_num_of_enrollment(row[5])
+            offload_or_recall_flag = get_offload_or_recall_flag(row[6])
 
+            # Input validation: if not valid input, reject the whole file and show error message with the incorrect column
+            if not is_valid_input(year, quarter, user_UCINetID_list, course_title_id, course_sec, num_of_enrollment, offload_or_recall_flag):
+                remove_upload_file(file)                    
+                return redirect(url_for('courses.offerings'))
 
-        # TODO: call "adjust_co_taught_course_pt": before checking if it's a co-taught course, 
-        #   I need to insert all data in the scheduled_teaching beforhand  
-        #   (rule: Points are divided equally between the instructors for a co-taught course.)
+            insert_or_update_scheduled_teaching_for_each_user(user_UCINetID_list, num_of_enrollment, user_id_and_academic_year_set, academic_year, combine_with, rows_dict, course_title_id, year, quarter, course_sec, offload_or_recall_flag, num_of_co_taught)            
 
-        # TODO: call calculate professor's point API (written by ying-ru): update_yearly_ending_balance(user_id, year, is_recursive) ??   (send academic year to ying-ru)
-                    
+        flash('Upload scheduled teaching file successfully!', 'success')
+
+        # rules #5: Combined grad/undergraduate classes
+        calculate_combined_classes_and_update_scheduled_teaching(rows_dict)        
```

Comment:
Somehow when I uploaded a new schedule_teaching_point file with a new enrollment number, I had an issue with the `TypeError: 'dict_keys' object is not subscriptable` error. Let's check this a bit offline.

Reply:
For `calculate_combined_classes_and_update_scheduled_teaching` function, would it call when`rows_dict` is not empty? So it is feasible to process keys from `rows_dict.keys()`.